File: 백준/Gold/2668. 숫자고르기/숫자고르기.java

import java.io.*;
import java.util.*;



public class Main {
    static boolean[] v;
    static int[] nums;
    static ArrayList<Integer> ans;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        nums = new int[(N+1)];
        for (int i = 1; i < N+1; i++) {
            nums[i] = Integer.parseInt(br.readLine());
        }

        ans = new ArrayList<>();

        for (int i = 1; i < N+1; i++) {
            v = new boolean[(N+1)];
            dfs(i,i);
        }

        System.out.println(ans.size());
        for (int a : ans) {
            System.out.println(a);
        }
    }

    public static void dfs(int start, int cur) {
        v[cur] = true;
        int next = nums[cur];

        if (!v[next]) {
            dfs(start,next);
        }

        else {
            if(start == next) {
                ans.add(start);
            }
        }

    }

}

File: 백준/Gold/2668. 숫자고르기/숫자고르기.py

def dfs(s,c) :

    v[c] =  1
    n = nums[c]

    if v[n] == 0 :
        dfs(s,n)
    else :
        if s == n :     # 시작점과 끝점이 같음 == 사이클 발견
            ans.add(s)


N = int(input())
nums = [0]+[int(input()) for _ in range(N)]
ans = set()

for i in range(1,N+1) :
    v = [0] * (N+1)
    dfs(i,i)

print(len(ans))
for a in ans :
    print(a)
